_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+    shift_and_log_scale_fn = tfb_masked_autoregressive_default_template(
+      hidden_layers = list(num_hidden, num_hidden),
+      activation = tf$nn$tanh))
+  bijectors[[i]] <- maf
+  if (use_batchnorm)
+    bijectors[[i + 1]] <- tfb_batch_normalization()
+  if (use_permute)
+    bijectors[[i + 2]] <- tfb_permute((ncol(df2) - 1):0)
+}
+
+if (use_permute) bijectors <- bijectors[-num_layers]
+
+flow <- bijectors %>%
+  discard(is.null) %>%
+  rev() %>%
+  tfb_chain()
+
+target_dist <- tfd_transformed_distribution(
+  distribution = base_dist,
+  bijector = flow
+)
+```
+
+And configuring the optimizer:
+
+```{r}
+optimizer <- tf$train$AdamOptimizer(1e-4)
+```
+
+Under that isotropic Gaussian we chose as a base distribution, how likely are the data?
+
+```{r}
+base_loglik <- base_dist %>% 
+  tfd_log_prob(x_train) %>% 
+  tf$reduce_mean()
+base_loglik %>% as.numeric()        # -11.33871
+
+base_loglik_test <- base_dist %>% 
+  tfd_log_prob(x_test) %>% 
+  tf$reduce_mean()
+base_loglik_test %>% as.numeric()   # -11.36431
+
+```
+
+
+And, just as a quick sanity check: What is the loglikelihood of the data under the transformed distribution _before any training_?
+
+```{r}
+target_loglik_pre <-
+  target_dist %>% tfd_log_prob(x_train) %>% tf$reduce_mean()
+target_loglik_pre %>% as.numeric()        # -11.22097
+
+target_loglik_pre_test <-
+  target_dist %>% tfd_log_prob(x_test) %>% tf$reduce_mean()
+target_loglik_pre_test %>% as.numeric()   # -11.36431
+```
+
+The values match - good. Here now is the training loop. Being impatient, here already we keep checking the loglikelihood on the (complete) test set to see if we're making any progress.
+
+```{r}
+n_epochs <- 10
+
+for (i in 1:n_epochs) {
+  
+  agg_loglik <- 0
+  num_batches <- 0
+  iter <- make_iterator_one_shot(train_dataset)
+  
+  until_out_of_range({
+    batch <- iterator_get_next(iter)
+    loss <-
+      function()
+        - tf$reduce_mean(target_dist %>% tfd_log_prob(batch))
+    optimizer$minimize(loss)
+    
+    loglik <- tf$reduce_mean(target_dist %>% tfd_log_prob(batch))
+    agg_loglik <- agg_loglik + loglik
+    num_batches <- num_batches + 1
+    
+    test_iter <- make_iterator_one_shot(test_dataset)
+    test_batch <- iterator_get_next(test_iter)
+    loglik_test_current <- target_dist %>% tfd_log_prob(test_batch) %>% tf$reduce_mean()
+    
+    if (num_batches %% 100 == 1)
+      cat(
+        "Epoch ",
+        i,
+        ": ",
+        "Batch ",
+        num_batches,
+        ": ",
+        (agg_loglik %>% as.numeric()) / num_batches,
+        " --- test: ",
+        loglik_test_current %>% as.numeric(),
+        "\n"
+      )
+  })
+}
+```
+
+With both training and test sets amounting to over 2 million records each, we did not have the patience to run this model _until no improvement occurred for 30 consecutive epochs on the validation set_. However, the picture we get from one complete epoch's run is pretty clear: The setup seems to work pretty okay.
```

Comment:
_until no improvement occurred for 30 consecutive epochs on the validation set_

^^ is this the specified norm from the paper? If so, you might say that in a parenthetical afterward.

Reply:
yeah I see the point, it's kind of a citation but, not formally... how about I add
(like the authors did)?

Reply:
"like" or "as", that's a difficult choice always ;-)

Reply:
I prefer as, but either way works!